aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <cstdint>
+#include <ATen/core/function_schema.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/TensorTypeSet.h>
+
+namespace c10 {
+
+/**
+ * An instance of DispatchKeyExtractor knows how to get a dispatch key given
+ * a list of arguments for an operator call. The instance is specific for
+ * a certain operator as different operators have different ways to extract
+ * the dispatch key (e.g. different numbers of arguments).
+ */
+struct DispatchKeyExtractor final {
+public:
+  static DispatchKeyExtractor make(const FunctionSchema& schema) {
+    bool is_valid = false;
+    for (size_t i = 0; i < schema.arguments().size(); ++i) {
+      const auto& type = schema.arguments()[i].type();
+      if (type->isSubtypeOf(TensorType::get())) {
+        is_valid = true;
+        break;
+      }
+      if (type->isSubtypeOf(ListType::ofTensors())) {
+        is_valid = true;
+        break;
+      }
+    }
+
+    return DispatchKeyExtractor(schema.arguments().size(), is_valid);
+  }
+
+  c10::optional<TensorTypeId> getDispatchKeyBoxed(const Stack* stack) const {
+    if (C10_UNLIKELY(!is_valid_)) {
```

Comment:
I don't want to address it in this PR because that would be a more intrusive change and I'm not sure I find time for it later, but I think it's a good idea.

Reply:
fixed